drivers/at86rf2xx: Fix L2 addr genration/handling

maribu opened the pull request:
### Contribution description

- Moved address generation to at86rf2xx_setup() to generate the L2 address only once
    - Resets will no longer change the L2 address
- Do not use uint16_t or uint64_t for short/long addresses to avoid confusion between host and network byte order
- No longer shuffling bytes of the LUID generated for use as L2 address:
    - Blindly setting/clearing the bits marking globally unique L2 addresses and multicast addresses no longer changes the byte luid_get() mutates, so no collisions of locally generated L2 addresses can occur any more
    - I still think it is a poor design decision to manipulate a LUID by hand and expect it to still be locally unique, but at least now it is (in the context of the current implementation of `luid_get()`) true

### Testing procedure

- Check if two notes with an `at86rf2xx` transceiver can still communicate
- Check if the remain able to communicate after multiple resets
- Check if two `at86rf2xx` transceivers (of the same type!) attached on the same board now get two different L2 addresses.
- Check if `ifconfig` still provides the same output except for the L2 addresses

### Issues/PRs references

Fixes https://github.com/RIOT-OS/RIOT/issues/9656

miri64 commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -56,6 +56,21 @@ void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
+     * to its desired value.
+     */
+    dev->netdev.long_addr[IEEE802154_LONG_ADDRESS_LEN - 1] &= ~(0x01);
+    dev->netdev.long_addr[IEEE802154_LONG_ADDRESS_LEN - 1] |=  (0x02);
```
This should be the _most_ significant byte, not the least. If you want to change the order internally however, you have to take this into account on `NETOPT` get/set where the order is specified as network byte-order (so msb first).

maribu replied:
I dropped the ntohl() **and** switched the byte to modify from the first to the last. That way the same byte is modified as before the PR on little endian systems. Either this is now still correct, or you just noticed another bug in the driver. I don't have a 802.15.4 spec at hands to verify. Do you mind to double check?

miri64 commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -100,14 +113,9 @@ void at86rf2xx_reset(at86rf2xx_t *dev)
-    at86rf2xx_set_addr_long(dev, ntohll(addr_long.uint64.u64));
-    at86rf2xx_set_addr_short(dev, ntohs(addr_long.uint16[0].u16));
+    at86rf2xx_set_addr_long(dev, dev->netdev.long_addr);
+    at86rf2xx_set_addr_short(dev, dev->netdev.short_addr);
```
Is it wise, just to assume the system is little-endian?

benpicco replied:
I would say so. RIOT does not support any big endian systems and that assumption is made in most places where RIOT talks to devices with little-endian multi-byte registers.

AFAIR there are no big endian MCUs.

maribu replied:
A byte array has no byte order. It makes no sense to convert byte order of a byte array.

miri64 replied:
True dat :sweat_smile:

miri64 commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -470,12 +470,12 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
         case NETOPT_ADDRESS:
             assert(len <= sizeof(uint16_t));
-            at86rf2xx_set_addr_short(dev, *((const uint16_t *)val));
+            at86rf2xx_set_addr_short(dev, val);
```
Here the order needs to be fixed for network byte-order though.

maribu replied:
Above you said that get()/set() is using network byte order for addresses, right?

miri64 replied:
Ok, for the short address I truly don't care, as it has no semantics to it.

miri64 commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -470,12 +470,12 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
         case NETOPT_ADDRESS_LONG:
             assert(len <= sizeof(uint64_t));
-            at86rf2xx_set_addr_long(dev, *((const uint64_t *)val));
+            at86rf2xx_set_addr_long(dev, val);
```
Same here.

miri64 commented:
>     * Check if `ifconfig` still provides the same output.

I added this test case to check if NETOPT is formatted properly.

maribu commented:
Note that opaque byte arrays received from luid_get are no longer shuffled arround, so on little endian systems the address will be "the other way round" (plus 4 bits changed). On the plus side, the same CPU ID will now generate the same L2 address regardless of endianess; which makes sense to me.

The change of the L2 addresses is pretty bad in OTA updates. But having in mind that a the way they were generated could result in e.g. four at86rf233 transceivers on the same board getting the exact same L2 addresses up to now, the generation of the addresses needed to be fixed. And I saw no value in having the L2 addresses being almost identical except for the colliding bits, so I optimised the byte order conversion of the byte array out. Maybe it is even better if the change in L2 addresses is more visible, than a sneaky change of only the colliding bits.

maribu commented:
Didn't we agree that byte arrays do not have a byte order?

miri64 commented:
> Didn't we agree that byte arrays do not have a byte order?

Yes, but it still matters which order you hand it up. The bits you are flipping after `luid_get()` are called the U/L bit and the G/L bit. They have significance in an EUI-64 and they have significance in the IID of an IPv6 address (where the U/L logic is flipped). If you put them now at the end of the byte array, and IPv6 and 6LoWPAN expect them at the beginning, this will lead problems we fought very hard in the past to get rid of.

miri64 commented:
Or to say it differently: by deciding to flip the bits of the last bytes you made your byte array an EUI-64 in little endian-order and the last byte the MSB. This isn't a problem as long as you just interact with the device this way. But when you hand up the long address to another layer it _will_ become a problem, as the EUI-64 is now in the wrong order.

See [IEEE guidelines of use of EUI](https://standards.ieee.org/content/dam/ieee-standards/standards/web/documents/tutorials/eui.pdf), [RFC 4291](https://tools.ietf.org/html/rfc4291#section-2.5.1), [RFC 4944](https://tools.ietf.org/html/rfc4944#section-8), [RFC 6775](https://tools.ietf.org/html/rfc6775#section-5.6),

maribu commented:
If there is no byte order in a byte array, it cannot be in the wrong order. So I assume what you are saying is, that I flipped the wrong bits in the address. But I didn't change the bits that were flipped.

Here the relevant part of the code in master:

``` C
     /* make sure we mark the address as non-multicast and not globally unique */ 
     addr_long.uint8[0] &= ~(0x01);                                               
     addr_long.uint8[0] |=  (0x02);                                               
    /* set short and long address */                                             
     at86rf2xx_set_addr_long(dev, ntohll(addr_long.uint64.u64));
```

With `ntohll` this will move the byte previously at index zero to index 7.

miri64 commented:
> Here the relevant part of the code in master:
> 
> ```c
>      /* make sure we mark the address as non-multicast and not globally unique */ 
>      addr_long.uint8[0] &= ~(0x01);                                               
>      addr_long.uint8[0] |=  (0x02);                                               
>     /* set short and long address */                                             
>      at86rf2xx_set_addr_long(dev, ntohll(addr_long.uint64.u64));
> ```
> 
> With `ntohll` this will move the byte previously at index zero to index 7.

Concerning this part of the code you are right, but at the line I commented in https://github.com/RIOT-OS/RIOT/pull/12600#discussion_r340235762, you hand the EUI-64 in little endian order (as the last byte got its bits flipped it became the MSB) up the stack, but the [doc of `NETOPT_ADDRESS_LONG`](https://doc.riot-os.org/group__net__netopt.html#gga19e30424c1ab107c9c84dc0cb29d9906a3f2a397a9a0dbf9673b3501566b05093) states it must be return in network byte order (aka big endian order).

miri64 commented on drivers/include/at86rf2xx.h:
```diff
@@ -36,6 +36,7 @@
 #include "net/netdev.h"
 #include "net/netdev/ieee802154.h"
 #include "net/gnrc/nettype.h"
+#include "net/eui64.h"
```
This include isn't needed btw.

benpicco commented:
Maybe the general confusion could be reduced by fixing the insanity contained inside the [at86rf2xx_set_addr_long()](https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_getset.c#L164) function…

maribu commented:
> Concerning this part of the code you are right

So the old driver generated the L2 address in big endian, converted it to little endian prior to the call in

> Maybe the general confusion could be reduced by fixing the insanity contained inside the [at86rf2xx_set_addr_long()](https://github.com/RIOT-OS/RIOT/blob/master/drivers/at86rf2xx/at86rf2xx_getset.c#L164) function…

Thought the same and did so! :laughing:

miri64 commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -56,6 +56,22 @@ void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
+         * transceivers are connected on the board.)
+         */
+    }
+    while ((dev->netdev.long_addr[0] & 0x03) == 0x02);
```
It was better before, now it makes completely no sense anymore.... first of all it wastes a lot of cycles + Now you are just checking if the bits at the first byte are set, but use it internally still in the order at hand (so the address is now used in the wrong order internally).

maribu replied:
OK. Let me explain the issue with just setting the bits of the first addresses.

Let's say the CPU ID is `0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`. In that case three consecutive calls to `luid_get()` with 8 bytes will return:

1. `0x01, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`
2. `0x02, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`
3. `0x03, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`

Those would now be changed to:

1. `0x02, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`
2. `0x02, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`
3. `0x02, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07`

So all three at86rf233 transceivers would gain the very same L2 address, which makes no sense.

maribu replied:
I fixed the first sentence in the reply above

benpicco replied:
At this point, why not just swap the first two bytes you get from `luid_get()`…?

maribu replied:
> At this point, why not just swap the first two bytes you get from `luid_get()`…?

Yes, a much better idea!

miri64 replied:
That's how it was originally...

maribu replied:
> That's how it was originally...

No. See: https://github.com/RIOT-OS/RIOT/blob/ea4b78654f45a33551d32ca205f35eac0aad4f56/drivers/at86rf2xx/at86rf2xx.c#L104-L107

maribu replied:
The current implementation changes the first byte returned by `luid_get()`. In context of the API of `luid_get()` every change of the returned locally unique ID breaks the guarantee of its uniqueness. In context of the current implementation, only the first byte must not be changed. The current code however does just this. Having two `at86rf233` transceivers on the same board with the current RIOT master will for some CPU IDs result in both transceivers having the same L2 address. That is one of the two bugs this PR tries to fix.

maribu replied:
This bug is also why one reset in https://github.com/RIOT-OS/RIOT/issues/9656 didn't break anything for the reporter, as that generated (due to that bug) the very same L2 address again. only after up to three resets a different L2 address is generated.

maribu replied:
Being honest: I still think that even with this PR applied the driver is still wrongly using `luid_get()`. It is totally insane to me to ask a unique 8 byte array, modify it, and still expect it to be unique. It only works now due to implementation details of `luid_get()`. Changing them would again break the L2 address generation.

If the standard use case of `luid` is to generate EUI64 and EUI48, maybe it is the time to add `luid_get_eui64()` and `luid_get_eui48()`? Those could enforce that the generated IDs are correctly marked as unicast and non-globally-unique and still unique for the boards they were generated. And this as part of the documentated guarantees of the API, not as a random internal implementation detail that might change. Also this would likely dedpulicate code.

benpicco replied:
> If the standard use case of luid is to generate EUI64 and EUI48, maybe it is the time to add `luid_get_eui64()` and `luid_get_eui48()`?

Yes, please!
A quick `grep` shows that *all* invocations but one of `luid_get()` are to set a MAC Address.
And everywhere those two magical bits are flipped.

If this discussion has has shown something, than it is that those functions are direly needed.

Also mind that all (?) drivers currently use the first two bytes of the long address for the short address, but I don't know if this is necessary. (I doubt so)

maribu commented:
> Concerning this part of the code you are right

So the old driver generated the L2 address in big endian, converted it to little endian prior to the call to `at86rf2xx_set_addr_long()`, in which it was converted back to big endian?

miri64 commented:
> > Concerning this part of the code you are right
> 
> So the old driver generated the L2 address in big endian, converted it to little endian prior to the call to `at86rf2xx_set_addr_long()`, in which it was converted back to big endian?

Yes... it's stupid.. That much we agree on.

miri64 commented:
The only thing I wanted from the original state of the PR is that this patch is applied

```
diff --git a/drivers/at86rf2xx/at86rf2xx_netdev.c b/drivers/at86rf2xx/at86rf2xx_netdev.c
index 0532c7571..133729cc2 100644
--- a/drivers/at86rf2xx/at86rf2xx_netdev.c
+++ b/drivers/at86rf2xx/at86rf2xx_netdev.c
@@ -473,11 +473,16 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
             at86rf2xx_set_addr_short(dev, val);
             /* don't set res to set netdev_ieee802154_t::short_addr */
             break;
-        case NETOPT_ADDRESS_LONG:
+        case NETOPT_ADDRESS_LONG: {
+            uint64_t addr;
+            const eui64_t *eui64 = val;
+
             assert(len <= sizeof(uint64_t));
-            at86rf2xx_set_addr_long(dev, val);
+            addr = byteorder_ntohll(eui64->uint64);
+            at86rf2xx_set_addr_long(dev, (uint64_t *)&addr);
             /* don't set res to set netdev_ieee802154_t::long_addr */
             break;
+        }
         case NETOPT_NID:
             assert(len <= sizeof(uint16_t));
             at86rf2xx_set_pan(dev, *((const uint16_t *)val));
```

and `dev->long_addr` is in the order expected by `netopt` and `netdev_ieee802154` -.-

miri64 commented:
I don't really care what happens internally in the device, but the network stack expects a long address where the U/L and G/L bits are in the first byte (when read as byte array). This needs to happen, everything else I don't care about ;-).

maribu commented:
If I understood you correct, the driver now sets the U/L and G/L bits correctly, right? So no need to shuffle bytes around any more, right?

maribu commented:
Sorry, forgot to push.

miri64 commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -150,23 +151,17 @@ void at86rf2xx_set_addr_short(at86rf2xx_t *dev, uint16_t addr)
+    for (unsigned i = 0; i < IEEE802154_LONG_ADDRESS_LEN; i++) {
         at86rf2xx_reg_write(dev, (AT86RF2XX_REG__IEEE_ADDR_0 + i),
-                            (addr >> ((7 - i) * 8)));
+                            dev->netdev.long_addr[i]);
```
Isn't this now in wrong order? The device expects the address to be LSB first, right?

miri64 replied:
I'm honestly unsure about this, the bit-shifting that is removed here makes things very much unclear ^^"

maribu replied:
Yes indeed. And I have been wrong about the IEEE address being a byte array. Maybe I should stick with the application layer. But I truly fail to understand, why is it was defined as a 64 bit number and not as 8 octets and therefore avoiding the confusion altogether...

maribu replied:
I'll test tomorrow to be sure pinging still works. It's too late for me to think straight.

miri64 replied:
> But I truly fail to understand, why is it was defined as a 64 bit number and not as 8 octets and therefore avoiding the confusion altogether...

Me neither, but apparently the original contributor thought it to be simple this way.

maribu commented:
I rebased on top of the new LUID API and on top of https://github.com/RIOT-OS/RIOT/pull/12606.

I would like to change the members `long_addr` and `short_addr` in [`netdev_ieee802154_t`](http://api.riot-os.org/structnetdev__ieee802154__t.html) to `eui64_t` and to `network_uint16_t` in a separate PR. This will make using the new LUID API mach easier.

Finally, I hope that this PR is only and intermediate step and will be replaced by the l2util @benpicco  is working on. But moving the network address retrieval from the reset procedure to the init function is something worth doing before that gets in.

fjmolinas commented:
@maribu what is the status with this PR? It doesn't seem to be waiting for any PR anymore.

maribu commented:
It currently is waiting for https://github.com/RIOT-OS/RIOT/pull/12713. Right now, `void luid_get_eui64(eui64_t *addr)` and `void luid_get_short(network_uint16_t *addr)` cannot be called with `uint8_t netdev_ieee802154_t::long_addr[8]` and `uint8_t netdev_ieee802154_t::short_addr[2]` as argument. Just casting would work, as `eui64_t` as an alignment requirement of 64 bit and `network_uint16_t` of 16 bit, while the data addresses in `netdev_ieee802154_t` only have an 8 bit alignment requirement.

I could work around this be assigning this to a temporary buffer and `memcpy()` it over. But that would be less readable and less efficient than just matching the type `luid_get_...()` expects and `netdev_ieee802154_t` uses.

I'm personally in favor of just consistently using `eui64_t`. I don't think that the fancy features of `eui64_t` are ever needed, though. So maybe a compromise would be to change `eui64_t` to make it less fancy:

```C
typedef struct {
    uint8_t value[8]; /**< EUI-64 in network byte order */
} eui64_t;
```

That would make the type applicable for low level stuff (as it is now a plain array of 8 bytes, just with a type name). And on the other hand it prevents the byte order confusion, as it is defined to be in network byte order.

fjmolinas commented:
@maribu how much of your test procedure still apply here?

fjmolinas commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -56,6 +56,9 @@ void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
     dev->params = *params;
 #endif
+    /* generate EUI-64 and short address */
+    luid_get_eui64(&dev->netdev.long_addr);
+    luid_get_short(&dev->netdev.short_addr);
```
```suggestion
    luid_get_eui64((eui64_t *) &dev->netdev.long_addr);
    luid_get_short((network_uint16_t *)&dev->netdev.short_addr);
```

maribu replied:
This won't work for now, as `eui64_t` has an alignment requirement of 64 bit, but `dev->netdev.long_addr` only has an alignment requirement of 8 bit. This cast would result in undefined behavior. E.g. on some ARM chips an unaligned access will result in hard faults..

The only options we have here are:

1. Use a temporary `eui64_t` and `memcpy()`, adding one unnecessary copy operation.
2. Change the API of `luid_get_eui64()` to accept a `void` Pointer as argument
3. Change the type of the member in `netdev_ieee802154_t::long_addr`.

(And the same for the short address.)

I don't think that we should go for 1. here. We have full control over our APIs and data types, so we should just make sure those are compatible with each other.

fjmolinas replied:
I'm sorry I misread what you had explained before, and though that your latest push was removing the parts of your PR that depended on additional changes.

maribu replied:
I did indeed remove some parts: In the meantime, `luid_get()` was replaced with `luid_get_eui64()`, fixing one bug this PR originally targeted. Now, only the reset issue (new address after each reset) is left.

fjmolinas replied:
Ok, I was naively suggesting the cast just because I had compiler errors.

fjmolinas replied:
Let try to reach a consensus in #12713 then

maribu replied:
Now that I'm aware that those crazy byteorder types use `__attribute__((packed))`, I'm fine with the cast. (There is not alignment mismatch due to this.)

fjmolinas commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -89,23 +92,16 @@ static void at86rf2xx_enable_smart_idle(at86rf2xx_t *dev)
     /* set short and long address */
-    at86rf2xx_set_addr_long(dev, &addr_long);
-    at86rf2xx_set_addr_short(dev, &addr_short);
+    at86rf2xx_set_addr_long(dev, dev->netdev.long_addr);
+    at86rf2xx_set_addr_short(dev, dev->netdev.short_addr);
```
```suggestion
    at86rf2xx_set_addr_long((network_uint16_t *)dev, dev->netdev.long_addr);
    at86rf2xx_set_addr_short((eui64_t *)dev, dev->netdev.short_addr);
```

fjmolinas replied:
I had initially put the cast wrong.

maribu commented:
I rebased to resolve merge conflicts and updated to use the common `ieee802154_setup()` function for the address generation as well. As the alignment mismatch between `uint8_t *` and `eui64_t *` is not actually present (`eui64_t` has an `__attribute__((packed))` buried somewhere in its definition), I just updated the code to cast between those.

maribu commented:
All green

maribu commented:
Thanks for the reviews :-)
